Formula/w/watchman.rb

```diff
@@ -62,7 +62,7 @@ def install
                     "-DPython3_EXECUTABLE=#{which("python3.12")}",
                     "-DWATCHMAN_VERSION_OVERRIDE=#{version}",
                     "-DWATCHMAN_BUILDINFO_OVERRIDE=#{tap.user}",
-                    "-DWATCHMAN_STATE_DIR=#{var}/run/watchman",
+                    "-DWATCHMAN_STATE_DIR=/tmp",
```

Comment:
I remember contemplating doing this to ensure portable bottles, but IIRC upstream strongly recommend against it. (They don't even set this for their own binary builds.)

We can, of course, ignore upstream, but it would be good to discuss it a little before we do that.

(Would provide more details but I'm about to catch a ✈️)

Reply:
@carlocab from a quick Google: Nix does this and it's working well for me locally.

What's the upstream concern? What do they set for their binaries?

Reply:
Possibly relevant upstream concern (and reasoning about why they didn't use `/tmp`) - https://github.com/facebook/watchman/issues/178#issuecomment-150783029.

> We changed the state dir to harden against a possible symlink attack (the state file, socket and log file could be redirected to some other location via a symlink to trick watchman into writing some place else).
> 
> It is important to have the sticky bit set on the statedir to help protect against this, which is why we didn't keep things in `/tmp`. You can change the statedir by doing `--enable-statedir=myprefix/foo` but you do need to keep in mind that there is a maximum practical length of approx 100 characters for unix domain socket names, so you don't want the path to be too deep.

Reply:
Any other suggested locations outside the prefix?

Reply:
`var/run` is pretty common for state files, so this may be a broader question: https://github.com/search?q=repo%3AHomebrew%2Fhomebrew-core+%22%7Bvar%7D%2Frun%22&type=code&p=2

Reply:
> `var/run` is pretty much standard for state files, so this may be a broader question: [github.com/search?q=repo%3AHomebrew%2Fhomebrew-core+%22%7Bvar%7D%2Frun%22&type=code&p=2](https://github.com/search?q=repo%3AHomebrew%2Fhomebrew-core+%22%7Bvar%7D%2Frun%22&type=code&p=2).

I don't think it is in this case given that it's not a permissions issue here, it's `watchman` itself refusing to run in this configuration (see error message above).

Honestly I'm not sure the security implications here outweigh the usability ones, particularly given the trust model on a typical Mac.

Reply:
Weird, I got a completely different error:

```
Uncaught C++ exception: std::__1::system_error: open: /opt/homebrew/var/run/watchman/bo-state: Permission denied
```

Reply:
Can someone either suggest an alternate location/fix or suggest how to move forward here?

The `watchman` formula is completely broken for me as-is. Note that the `/tmp` I suggested is also what Nix does: https://github.com/NixOS/nixpkgs/blob/dcfcc14e974d5efb9916433cdfb13f6f9cd50cc6/pkgs/development/tools/watchman/default.nix#L28

Reply:
Looks like `--disable-statedir` means that it will use `$TMPDIR` which is significantly more secure (on macOS at least):

```
$ stat -f '%Sp %Su %Sg' /tmp   
lrwxr-xr-x root wheel
$ stat -f '%Sp %Su %Sg' $TMPDIR
drwx------ bo staff
```

https://github.com/facebook/watchman/blob/7ac85af16a1d625dccf4e94619f889d6fa902fc1/watchman/UserDir.cpp#L148-L152
https://github.com/facebook/watchman/blob/7ac85af16a1d625dccf4e94619f889d6fa902fc1/watchman/UserDir.cpp#L82

Reply:
@bo98 great find, thanks!

Reply:
> I think a blank `-DWATCHMAN_STATE_DIR=` might do that? Haven't tested but I think that's righ

Have pushed a chance for this and will try it locally, too.